cupy/_core/core.pyx

```diff
@@ -342,6 +354,44 @@ cdef class _ndarray_base:
             device_type = dlpack.device_ROCM
         return (device_type, self.device.id)
 
+    def __getbuffer__(self, Py_buffer* buf, int flags):
+        # TODO(leofang): use flags
+        if (not is_sam_supported(self.data.device_id)
+                or not self.is_host_accessible()):
+            raise RuntimeError(
+                'Accessing a CuPy ndarry on CPU is not allowed except when '
+                'using system memory (on HMM or ATS enabled systems, need to '
+                'set CUPY_ENABLE_SAM=1) or managed memory')
+
```

Comment:
Note: I had a TODO here mentioning the possibility of prefetching. However, it seems prefetching is slow and I did not have time to build a proper heuristic for prefetching.